caikit/core/data_model/dataobject.py

```diff
@@ -18,99 +18,64 @@
+from typing import Any, Callable, List, Type, Union, get_args, get_origin
+import dataclasses
 import importlib
 import sys
 import types
-import typing
 
 # Third Party
-from google.protobuf import descriptor as _descriptor
 from google.protobuf import message as _message
 from google.protobuf.internal.enum_type_wrapper import EnumTypeWrapper
 
 # First Party
-from jtd_to_proto.jtd_to_proto import JTD_TO_PROTO_TYPES
-from jtd_to_proto.validation import is_valid_jtd
+from py_to_proto.dataclass_to_proto import DataclassConverter
 import alog
-import jtd_to_proto
+import py_to_proto
 
 # Local
 from ..toolkit.errors import error_handler
 from . import enums
 from .base import DataBase, _DataBaseMetaClass
-from .streams.data_stream import DataStream
 
 ## Globals #####################################################################
 
 log = alog.use_channel("SCHEMA")
 error = error_handler.get(log)
 
-# Type defs for input schemas to a dataobject
-_SCHEMA_VALUE_TYPE = Union[str, "_SCHEMA_VALUE_TYPE", Type[DataBase]]
-_SCHEMA_DEF_TYPE = Dict[str, _SCHEMA_VALUE_TYPE]
-
 # Registry of auto-generated protos so that they can be rendered to .proto
 _AUTO_GEN_PROTO_CLASSES = []
 
-# Reserved keywords in JTD
-_JTD_KEYWORDS = [
-    "elements",
-    "type",
-    "properties",
-    "enum",
-    "values",
-    "optionalProperties",
-    "additionalProperties",
-    "discriminator",
-]
-
-# Type defs for schemas passed to jtd_to_proto
-_JTD_VALUE_TYPE = Union[str, "_JTD_VALUE_TYPE", _descriptor.Descriptor]
-_JTD_DEF_TYPE = Dict[str, _JTD_VALUE_TYPE]
-
-# Python type -> jtd name
-_NATIVE_TYPE_TO_JTD = {
-    str: "string",
-    int: "int64",
-    float: "float64",
-    bytes: "bytes",
-    bool: "boolean",
-    datetime: "timestamp",
-}
+# Special attribute used to indicate which defaults are user provided
+_USER_DEFINED_DEFAULTS = "__user_defined_defaults__"
 
 ## Public ######################################################################
 
 # Common package prefix
 CAIKIT_DATA_MODEL = "caikit_data_model"
 
 
-def dataobject(
-    schema: _SCHEMA_DEF_TYPE,
-    package: str = CAIKIT_DATA_MODEL,
-) -> Callable[[Type], Type[DataBase]]:
-    """The @schema decorator can be used to define a Data Model object's schema
-    inline with the definition of the python class rather than needing to bind
-    to a pre-compiled protobufs class. For example:
+def dataobject(*args, **kwargs) -> Callable[[Type], Type[DataBase]]:
+    """The @dataobject decorator can be used to define a Data Model object's
+    schema inline with the definition of the python class rather than needing to
+    bind to a pre-compiled protobufs class. For example:
 
-    @dataobject(
-        package="foo.bar",
-        schema={"foo": str, "bar": int},
-    )
+    @dataobject("foo.bar")
+    @dataclass
     class MyDataObject:
         '''My Custom Data Object'''
+        foo: str
+        bar: int
 
     NOTE: The wrapped class must NOT inherit directly from DataBase. That
```

Comment:
Just a question: Is there any way to get that to work?

This snippet makes a static type checker barf but runtime type checking passes:
```python
@dataobject
@dataclass
class MyDataObject:
    foo: int

def my_func(data_model: DataBase):
    assert isinstance(data_model, DataBase)

my_func(MyDataObject())
```